src/mumble/Settings.h

```diff
@@ -431,10 +431,14 @@ struct Settings {
 	QByteArray qbaPTTButtonWindowGeometry = {};
 
 	// Network settings
-	bool bTCPCompat   = false;
-	bool bReconnect   = true;
-	bool bAutoConnect = false;
-	bool bQoS         = true;
+	bool bUdpDisabled            = false;
+	bool bUdpStreamInboundOnly   = false;
+	bool bUdpStreamOutboundOnly  = false;
+	bool bUdpStreamBidirectional = false;
```

Comment:
I'm wondering whether this might be better served by a single variable of `enum` type:
```cpp
enum class UDPMode {
	Disabled,
	Inbound,
	Outbound,
	Bidirectional,
};

UDPMode udpMode = UDPMode::Bidirectional;
```
or maybe even as bit-flags:
```cpp
namespace UDPMode {
	enum Mode {
		Disabled = 0,
		Inbound = 1 << 0,
		Outbound = 1 << 1,
		Bidirectional = Inbound | Outbound,
	};
}

UDPMode::Mode udpMode = UDPMode::Bidirectional
```
and then check like this:
```cpp
if (udpMode & UDPMode::Inbound) {
	// Do stuff
}
````
This has the advantage that above code works as expected if `udpMode` is `Bidirectional` without explicitly checking `udpMode == Inbound || udpMode == Bidirectional` (which is what would be needed with my first approach)

Reply:
The bit-flags approach seems to be better. I will go ahead with that.

Reply:
Is there any special advantage of using the namespace part? Can I use `enum UDPMode` directly within the class 'Settings'?
Because I see that the enum type's scope will still be inside the class 'Settings'. Or maybe it is just a sample code to make me understand what changes to make here?
Please let me know.

Reply:
```
PROCESS(network, RESTRICT_TO_TCP_KEY, bUdpDisabled)                      \
PROCESS(network, UDP_INBOUND_ONLY_KEY, bUdpStreamInboundOnly)            \
PROCESS(network, UDP_OUTBOUND_ONLY_KEY, bUdpStreamOutboundOnly)          \
PROCESS(network, UDP_BIDIRECTIONAL_KEY, bUdpStreamBidirectional)         \
PROCESS(network, UDP_AUTO_KEY, bUdpStreamAuto)                           \
```

Earlier I was using bool variable for each UDP mode so it was straight forward to map them as shown in the above code. But now with a single enum variable I have no idea how to put it there. Can you please help me find a way to do it?

Reply:
> So by the namespace->enum suggestion you mean the definition of the enum is in the global scope? Not inside 'Settings' class?

Yes. I'm not a big fan of having the enum definitions in the `Settings` namespace as they are used also (actually, mainly) outside a settings context.

> And what about the UDPMode::Auto?

That's a different context I would say. While the first 4 states (Disabled, InboundOnly, OutboundOnly, Both) describe the currently used state of the UDP direction, the `Auto` mode determines if and how Mumble should switch between these states.
To be more concrete: If at some point we need to check whether we want to send out audio via UDP, we check if we're in `OutboundOnly` or `Both` mode. However, we don't care in the slightest whether we're in `Auto` mode. The latter only comes into play, once we figure that the UDP packets that we send out don't reach their destination.
Thus, we are expressing different concepts here, which is why I would not want to fuse that into a single type.

> Earlier I was using bool variable for each UDP mode so it was straight forward to map them as shown in the above code. But now with a single enum variable I have no idea how to put it there. Can you please help me find a way to do it?

In order for this trick to work with enums, we require a bit more boilerplate code for the string <-> enum conversion. This is done in the `EnumStringConversion` namespace and you will find examples for other enums there. Once you have defined the necessary functions for the new enum, you should be able to use the usual mechanism for storing and saving the settings.

For representing bit-flags, I think we could leverage a symbolic notation like this:
```cpp
#define UDP_MODES                                    \
	PROCESS(UDPMode::Mode, Disabled, "Disabled")     \
	PROCESS(UDPMode::Modet, InboundOny, "Inbound")   \
	PROCESS(UDPMode::Mode, OutboundOnly, "Outbound") \
	PROCESS(UDPMode::Mode, Both, "Inbound | Outbound")
```

Reply:
> That's a different context I would say. While the first 4 states (Disabled, InboundOnly, OutboundOnly, Both) describe the currently used state of the UDP direction, the `Auto` mode determines if and how Mumble should switch between these states. To be more concrete: If at some point we need to check whether we want to send out audio via UDP, we check if we're in `OutboundOnly` or `Both` mode. However, we don't care in the slightest whether we're in `Auto` mode. The latter only comes into play, once we figure that the UDP packets that we send out don't reach their destination. Thus, we are expressing different concepts here, which is why I would not want to fuse that into a single type.

So that means the **AUTO** is meant to be an additional advanced feature of the UDP mode feature not part of this commit. I am guessing from your first comment on issue [#5517](https://github.com/mumble-voip/mumble/issues/5517).

Reply:
Well, it could be added later in a separate commit, but if you have an idea on how to make it work, I also don't object to including it in this one. That's up to you.